From 09d598521feeabf2d468cfff08461610d639697f Mon Sep 17 00:00:00 2001
From: seekilm <seekilm92@gmail.com>
Date: Wed, 19 Feb 2025 12:04:45 +0100
Subject: [PATCH] dwm: add support for Belgian keyboard (be-latin1)

---
 config.def.h | 18 +++++++++---------
 1 file changed, 9 insertions(+), 9 deletions(-)

diff --git a/config.def.h b/config.def.h
index 9efa774..fa0685b 100644
--- a/config.def.h
+++ b/config.def.h
@@ -85,15 +85,15 @@ static const Key keys[] = {
 	{ MODKEY,                       XK_period, focusmon,       {.i = +1 } },
 	{ MODKEY|ShiftMask,             XK_comma,  tagmon,         {.i = -1 } },
 	{ MODKEY|ShiftMask,             XK_period, tagmon,         {.i = +1 } },
-	TAGKEYS(                        XK_1,                      0)
-	TAGKEYS(                        XK_2,                      1)
-	TAGKEYS(                        XK_3,                      2)
-	TAGKEYS(                        XK_4,                      3)
-	TAGKEYS(                        XK_5,                      4)
-	TAGKEYS(                        XK_6,                      5)
-	TAGKEYS(                        XK_7,                      6)
-	TAGKEYS(                        XK_8,                      7)
-	TAGKEYS(                        XK_9,                      8)
+	TAGKEYS(                        XK_ampersand,              0)
+	TAGKEYS(                        XK_eacute,                 1)
+	TAGKEYS(                        XK_quotedbl,               2)
+	TAGKEYS(                        XK_apostrophe,             3)
+	TAGKEYS(                        XK_parenleft,              4)
+	TAGKEYS(                        XK_section,                5)
+	TAGKEYS(                        XK_egrave,                 6)
+	TAGKEYS(                        XK_exclam,		   7)
+	TAGKEYS(                        XK_ccedilla,               8)
 	{ MODKEY|ShiftMask,             XK_q,      quit,           {0} },
 };
 
-- 
2.48.1

